### PR TITLE
[improve][admin,broker] Add option to  unloadNamespaceBundle with bundle Affinity broker url

### PR DESCRIPTION
**Pull request by vineeth1995**
### Motivation

To provide an option for admin to unload namespace bundle with broker url to which the bundle has to be assigned to after the unload.

### Modifications

1. Add an option to provide brokerWebServiceAddress (Ex: https://broker1.com:4443) while unloadingNamespaceBundle in the admin CLI and make corresponding changes in the pulsar-client-admin-api to support this.

2. Changes in pulsar-broker unloadNamespaceBundle API to accommodate  broker url while unloading. 

3. Add functionality in LoadManger implementation to select affinity broker for a bundle while trying to assign unloaded bundle to broker.

### Verifying this change

This change added tests and can be verified as follows:

  - Added end to test to make sure the bundle assignment happens as per the expectation

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [x] The REST endpoints
- [x] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 1]
```diff
@@ -853,6 +854,28 @@ protected BookieAffinityGroupData internalGetBookieAffinityGroup() {
...
     }
 
+    public void setNamespaceBundleAffinity (String bundleRange, String brokerUrl) {
+        if (brokerUrl != null) {
+            if (!this.isLeaderBroker()) {
+                LeaderBroker leaderBroker = pulsar().getLeaderElectionService().getCurrentLeader().get();
+                String leaderBrokerUrl = leaderBroker.getServiceUrl();
+                try {
+                    URL redirectUrl = new URL(leaderBrokerUrl);
+                    URI redirect = UriBuilder.fromUri(uri.getRequestUri()).host(redirectUrl.getHost())
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:863]: we need to return `http` or `https` based on user request. Just check existing code how it performs redirection.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 2]
```diff
@@ -853,6 +854,28 @@ protected BookieAffinityGroupData internalGetBookieAffinityGroup() {
...
+                            .port(redirectUrl.getPort()).replaceQueryParam("authoritative",
+                                    false).build();
+
+                    // Redirect
+                    log.debug("Redirecting the rest call to {}", redirect);
+                    throw new WebApplicationException(Response.temporaryRedirect(redirect).build());
+                } catch (MalformedURLException e) {
+                    e.printStackTrace();
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java]: printstackTrace is expensive as it performs IO. Never use printstackTrace or Sysout in prod. Always use logger.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 3]
```diff
@@ -853,6 +854,28 @@ protected BookieAffinityGroupData internalGetBookieAffinityGroup() {
...
+                LeaderBroker leaderBroker = pulsar().getLeaderElectionService().getCurrentLeader().get();
+                String leaderBrokerUrl = leaderBroker.getServiceUrl();
+                try {
+                    URL redirectUrl = new URL(leaderBrokerUrl);
+                    URI redirect = UriBuilder.fromUri(uri.getRequestUri()).host(redirectUrl.getHost())
+                            .port(redirectUrl.getPort()).replaceQueryParam("authoritative",
+                                    false).build();
+
+                    // Redirect
+                    log.debug("Redirecting the rest call to {}", redirect);
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java]: it should have more logging param : topic name at least and `debug` log should go under `if` statement. check debug log statement.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java [hunk 4]
```diff
@@ -885,8 +885,10 @@ public void unloadNamespace(@Suspended final AsyncResponse asyncResponse, @PathP
     public void unloadNamespaceBundle(@Suspended final AsyncResponse asyncResponse,
             @PathParam("property") String property, @PathParam("cluster") String cluster,
             @PathParam("namespace") String namespace, @PathParam("bundle") String bundleRange,
-            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative) {
+            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
+            @QueryParam("brokerUrl") String brokerUrl) {
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/Namespaces.java]: can we rename as `destinationBroker`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java [hunk 5]
```diff
@@ -812,8 +812,10 @@ public void unloadNamespace(@Suspended final AsyncResponse asyncResponse,
     public void unloadNamespaceBundle(@Suspended final AsyncResponse asyncResponse,
             @PathParam("tenant") String tenant, @PathParam("namespace") String namespace,
             @PathParam("bundle") String bundleRange,
-            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative) {
+            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
+                                      @QueryParam("brokerUrl") String brokerUrl) {
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java]: same here `destinationbroker`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java [hunk 6]
```diff
@@ -197,6 +197,7 @@ public class ModularLoadManagerImpl implements ModularLoadManager {
 
     private final Lock lock = new ReentrantLock();
     private Set<String> knownBrokers = ConcurrentHashMap.newKeySet();
+    private ConcurrentHashMap<String, String> bundleBrokerAffinityMap;
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java]: It should be `Map`<>` but when you initialize, you can use specific Map impl class and In this case `ConcurrentHashMap`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerWrapper.java [hunk 7]
```diff
@@ -65,13 +65,14 @@ public LoadManagerReport generateLoadReport() {
 
     @Override
     public Optional<ResourceUnit> getLeastLoaded(final ServiceUnitId serviceUnit) {
+        String bundleRange = LoadManagerShared.getBundleRangeFromBundleName(serviceUnit.toString());
+        String affinityBroker = loadManager.getNamespaceBundleAffinity(bundleRange);
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerWrapper.java]: can we just do `remove(..)` as `remove` returns the value and also removes the key.

**Comment by rdhabalia** [comment]
it seems multiple test cases are failing with this change
```
  [INFO] Running org.apache.pulsar.broker.namespace.NamespaceCreateBundlesTest
  Error:  Tests run: 6, Failures: 2, Errors: 0, Skipped: 4, Time elapsed: 0.059 s <<< FAILURE! - in org.apache.pulsar.broker.namespace.NamespaceCreateBundlesTest
  Error:  testCreateNamespaceWithDefaultBundles(org.apache.pulsar.broker.namespace.NamespaceCreateBundlesTest)  Time elapsed: 0 s  <<< FAILURE!
  org.apache.pulsar.tests.FailFastNotifier$FailFastSkipException: Skipped after failure since testFailFast system property is set.
  	at org.apache.pulsar.tests.FailFastNotifier.beforeInvocation(FailFastNotifier.java:88)
  
  Error:  testSplitBundleUpdatesLocalPoliciesWithoutOverwriting(org.apache.pulsar.broker.namespace.NamespaceCreateBundlesTest)  Time elapsed: 0.003 s  <<< FAILURE!
  org.apache.pulsar.tests.FailFastNotifier$FailFastSkipException: Skipped after failure since testFailFast system property is set.
  	at org.apache.pulsar.tests.FailFastNotifier.beforeInvocation(FailFastNotifier.java:88)
  
  [INFO] Running org.apache.pulsar.broker.namespace.NamespaceServiceTest
  [INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.077 s - in org.apache.pulsar.broker.namespace.NamespaceServiceTest
  [INFO] Running org.apache.pulsar.broker.namespace.OwnershipCacheTest
  Error:  Tests run: 16, Failures: 8, Errors: 0, Skipped: 8, Time elapsed: 0.075 s <<< FAILURE! - in org.apache.pulsar.broker.namespace.OwnershipCacheTest
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java (hunk 1 above)
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:863]: This will do the same based on http or https

**Pull request closed without merging**